topology-manager/src/main/java/org/apache/aries/rsa/topologymanager/importer/MultiMap.java

```diff
@@ -19,60 +19,54 @@
 package org.apache.aries.rsa.topologymanager.importer;
 
 import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
 
 /**
- * Minimal implementation of a synchronized map
+ * Minimal implementation of a thread-safe map where each key can have multiple values.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
  */
-public class MultiMap<T> {
+public class MultiMap<K, V> {
 
-    private Map<String, Set<T>> map;
+    private Map<K, Set<V>> map;
     
     public MultiMap() {
-        map = new HashMap<>();
-    }
-    
-    public synchronized void put(String key, T value) {
-        Set<T> values = map.get(key);
-        if (values == null) {
-            values = new HashSet<>();
-            map.put(key, values);
-        }
-        values.add(value);
+        map = new ConcurrentHashMap<>();
     }
     
-    public synchronized Set<T> get(String key) {
-        if (map.containsKey(key)) {
-            return Collections.unmodifiableSet(new HashSet<>(map.get(key)));
-        } else {
-            return Collections.<T>emptySet();
-        }
+    public void put(K key, V value) {
+        map.compute(key, (k, v) -> {
+            if (v == null)
+                v = new CopyOnWriteArraySet<>();
+            v.add(value);
+            return v;
+        });
     }
 
-    public synchronized void remove(String key, T value) {
-        Set<T> values = map.get(key);
-        values.remove(value);
-        if (values.isEmpty()) {
-            map.remove(key);
-        }
+    // the returned set should not be modified directly
+    public Set<V> get(K key) {
+        return map.getOrDefault(key, Collections.emptySet());
     }
 
-    public synchronized Set<String> keySet() {
-        return Collections.unmodifiableSet(new HashSet<>(map.keySet()));
+    public void remove(K key, V value) {
+        map.compute(key, (k, v) -> v != null && v.remove(value) && v.isEmpty() ? null : v);
```

Comment:
I would extend this to a multi-line lambda to improve readability.

Reply:
Not sure how you would expand this... seems pretty straight-forward to me ;-)
If it helps I can add a comment reminding that returning null will remove the mapping.

Reply:
It is a matter of taste I guess :). The null returning is something that I had to look up in the Javadoc, I may not be alone in learning a new trick here :)

Reply:
Added a clarifying comment.